Tutorial: Get started with k6 (#1192)
grafana/k6-docs: src/data/markdown/docs/05 Examples/02 Tutorials/01 Get started with k6/200 Test for performance.md

```
✓ http_req_duration..............: avg=66.14ms    min=0s         med=0s         max=198.42ms   p(90)=158.73ms   p(95)=178.58ms
       { expected_response:true }...: avg=198.42ms   min=198.42ms   med=198.42ms   max=198.42ms   p(90)=198.42ms   p(95)=198.42ms
✗ http_req_failed................: 0.00% ✓ 0        ✗ 1
```

## Test performance under increasing load

Now your script has logic to simulate user behavior, and assertions for functionality (checks) and performance (thresholds).

It's time to increase load to see how it performs.
To increase load, use the scenarios property.
Scenarios schedule load according to number of VUs, number of iterations, VUs, or by iteration rate.

### Run a smoke test

Start small. Run a [smoke test](/test-types/smoke-testing "a small test to confirm the script works properly") to see your script can handle minimal load.

To do so, use the [`--iterations`](/using-k6/k6-options/reference/#iterations) flag with an argument of 10 or fewer.

```bash
k6 run --iterations 10 api-test.js
```

If the service can't receive 10 iterations, the system has some serious performance issues to debug.
Good thing you ran the test early!

> ### Run a test against average load

Reviewer: ```suggestion
### Run a test against an average load
```

Now increase load.

Generally, traffic doesn't arrive all at once.
Rather, it gradually increases to a peak load.
To simulate this, testers increase load in _stages_.

Since this is a learning environment, the stages are still quite short.

> Add the following _scenario_ to your options `object` and run the test again.

Reviewer: ```suggestion
Add the following _scenario_ to your options `object` and rerun the test.
```

```json
  scenarios: {
    //arbitrary name of scenario
    average_load: {
      executor: "ramping-vus",
      stages: [
        // ramp up to average load of 20 virtual users
        { duration: "10s", target: 20 },
        // maintain load
        { duration: "50s", target: 20 },
        // ramp down to zero
        { duration: "5s", target: 0 },
      ],
    },
  }
```

Run the test with no command-line flags: 

```bash
k6 run api-test.js
```

The load is small, so the server should perform within thresholds.
However, this test server may be under load by many k6 learners, so the results are unpredictable.

<Blockquote mod="note" title="To visualize results...">

At this point, it'd be nice to have a graphical interface to visualize metrics as they occur.
k6 has many output formats, which can serve as inputs for many visualization tools, both open source and commercial.
For ideas, read [Ways to visualize k6 results](https://k6.io/blog/ways-to-visualize-k6-results).

</Blockquote>

### Ramp up until threshold fails

> Finally, run a [breakpoint test](/test-types/breakpoint-testing), where you probe the limits of the system.

Reviewer: ```suggestion
Finally, run a [breakpoint test](/test-types/breakpoint-testing), where you probe the system's limits.
```

In this case, run the test until the availability (error rate) threshold is crossed.

To do this:

> 1. Configure the threshold so that it aborts when it fails.

Reviewer: ```suggestion
1. Configure the threshold to abort when it fails.
```

  ```javascript
  http_req_failed: [{ threshold: "rate<0.01", abortOnFail: true }], // http errors should be less than 1%, otherwise abort the test
  ```

1. Configure the load to ramp the test up until it fails.

  ```javascript
  export const options = {
    thresholds: {
      http_req_failed: [{ threshold: "rate<0.01", abortOnFail: true }], 
      http_req_duration: ['p(99)<1000'], 
    },
    scenarios: {
      //arbitrary name of scenario:
      breaking: {
        executor: "ramping-vus",
        stages: [
          { duration: "10s", target: 20 },
          { duration: "50s", target: 20 },
          { duration: "50s", target: 40 },
          { duration: "50s", target: 60 },
          { duration: "50s", target: 80 },
          { duration: "50s", target: 100 },
          { duration: "50s", target: 120 },
          { duration: "50s", target: 140 },
          //....
        ],
      },
    },
  }
  ```

Here is the full script.

> Copy and run it.

Reviewer: ```suggestion
Copy and run it with `k6 run api-test.js`.
```
